File: frontend/src/servicios/traspasos.ts
/**
 * Cliente de traspasos entre sucursales (US6). Despachar deja la mercancía en tránsito;
 * confirmar la recepción expone la discrepancia por producto, sin clasificarla.
 */

import { clienteHttp } from "./clienteHttp";

export interface RenglonTraspaso {
  id_producto: number;
  cantidad_despachada: number;
  cantidad_recibida: number | null;
  discrepancia: number | null;
}

export interface Traspaso {
  id_traspaso: number;
  id_sucursal_origen: number;
  id_sucursal_destino: number;
  estado: "en_transito" | "recibido";
  instante_despacho: string;
  instante_recepcion: string | null;
  renglones: RenglonTraspaso[];
}

export function despacharTraspaso(datos: {
  id_sucursal_origen: number;
  id_sucursal_destino: number;
  renglones: { id_producto: number; cantidad: number }[];
}): Promise<Traspaso> {
  return clienteHttp.post<Traspaso>("/traspasos", datos);
}

export function listarTraspasos(opciones: {
  estado?: Traspaso["estado"];
  idSucursalOrigen?: number;
  idSucursalDestino?: number;
} = {}): Promise<Traspaso[]> {
  const params = new URLSearchParams();
  if (opciones.estado) params.set("estado", opciones.estado);
  if (opciones.idSucursalOrigen !== undefined)
    params.set("id_sucursal_origen", String(opciones.idSucursalOrigen));
  if (opciones.idSucursalDestino !== undefined)
    params.set("id_sucursal_destino", String(opciones.idSucursalDestino));
  const q = params.toString();
  return clienteHttp.get<Traspaso[]>(`/traspasos${q ? `?${q}` : ""}`);
}

export function recibirTraspaso(
  idTraspaso: number,
  renglones: { id_producto: number; cantidad_recibida: number }[],
): Promise<Traspaso> {
  return clienteHttp.post<Traspaso>(`/traspasos/${idTraspaso}/recepcion`, { renglones });
}

File: backend/rasero/api/traspasos.py
"""Endpoints de traspaso entre sucursales (T076, T077, US6):

- `POST /traspasos` — despachar: mercancía a tránsito, sigue en el inventario total (FR-018).
- `POST /traspasos/{id_traspaso}/recepcion` — confirmar: entrada en destino, expone la
  discrepancia por producto sin clasificarla (FR-019).
"""

from fastapi import APIRouter, Depends, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from rasero.persistencia.sesion import obtener_sesion
from rasero.seguridad import exige_rol
from rasero.servicios import traspasos as servicio_traspasos

# Traspasos entre sucursales: rol `encargado` (constitución v2.5.0, "Autorización de pantalla").
# Tanto el despacho como la recepción — la pantalla de recepción sólo se alcanza desde la de
# despacho, no hay flujo de `cajero` que traspase mercancía.
router = APIRouter(tags=["traspasos"], dependencies=[Depends(exige_rol("encargado"))])


class RenglonDespachoNuevo(BaseModel):
    id_producto: int
    cantidad: int = Field(ge=1)


class TraspasoNuevo(BaseModel):
    id_sucursal_origen: int
    id_sucursal_destino: int
    renglones: list[RenglonDespachoNuevo] = Field(min_length=1)


class RenglonRecepcionNuevo(BaseModel):
    id_producto: int
    cantidad_recibida: int = Field(ge=0)


class RecepcionTraspaso(BaseModel):
    renglones: list[RenglonRecepcionNuevo] = Field(min_length=1)


@router.get("/traspasos")
def listar_traspasos(
    estado: str | None = Query(default=None),
    id_sucursal_origen: int | None = Query(default=None),
    id_sucursal_destino: int | None = Query(default=None),
    sesion: Session = Depends(obtener_sesion),
) -> list[dict]:
    return servicio_traspasos.listar_traspasos(
        sesion,
        estado=estado,
        id_sucursal_origen=id_sucursal_origen,
        id_sucursal_destino=id_sucursal_destino,
    )


@router.post("/traspasos", status_code=status.HTTP_201_CREATED)
def despachar_traspaso(cuerpo: TraspasoNuevo, sesion: Session = Depends(obtener_sesion)) -> dict:
    traspaso = servicio_traspasos.despachar_traspaso(
        sesion,
        id_sucursal_origen=cuerpo.id_sucursal_origen,
        id_sucursal_destino=cuerpo.id_sucursal_destino,
        renglones=[
            servicio_traspasos.RenglonDespacho(id_producto=r.id_producto, cantidad=r.cantidad)
            for r in cuerpo.renglones
        ],
    )
    return servicio_traspasos.traspaso_a_respuesta(sesion, traspaso)


@router.post("/traspasos/{id_traspaso}/recepcion")
def recibir_traspaso(
    id_traspaso: int, cuerpo: RecepcionTraspaso, sesion: Session = Depends(obtener_sesion)
) -> dict:
    traspaso = servicio_traspasos.recibir_traspaso(
        sesion,
        id_traspaso=id_traspaso,
        renglones=[
            servicio_traspasos.RenglonRecepcion(
                id_producto=r.id_producto, cantidad_recibida=r.cantidad_recibida
            )
            for r in cuerpo.renglones
        ],
    )
    return servicio_traspasos.traspaso_a_respuesta(sesion, traspaso)
